flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java

```diff
@@ -742,13 +746,24 @@ public void goToFailing(
                         failureCause));
     }
 
+    @Override
+    public int getAvailableSlots() {
+        return declarativeSlotPool.getFreeSlotsInformation().size();
+    }
+
+    @Override
+    public Optional<? extends VertexAssignment> getVertexAssignmentWithAllSlots() {
```

Comment:
Wildcards in return types are discouraged because they only make the API less comfortable. Please return `Optional<VertexAssignment>`.